services/kvstore/src/main/java/org/hyperledger/besu/services/kvstore/SegmentedInMemoryKeyValueStorage.java

```diff
@@ -143,27 +145,70 @@ public Optional<byte[]> get(final SegmentIdentifier segmentIdentifier, final byt
+        e ->
+            compareKeyLeftToRight(e.getKey(), key) <= 0
+                && e.getKey().commonPrefixLength(key) >= e.getKey().size(),
+        e -> compareKeyLeftToRight(e.getKey(), key) < 0,
+        false);
+  }
+
+  @Override
+  public Optional<NearestKeyValue> getNearestAfter(
+      final SegmentIdentifier segmentIdentifier, final Bytes key) throws StorageException {
+    return getNearest(
+        segmentIdentifier,
+        e ->
+            compareKeyLeftToRight(e.getKey(), key) >= 0
+                && e.getKey().commonPrefixLength(key) >= e.getKey().size(),
+        e -> compareKeyLeftToRight(e.getKey(), key) >= 0,
+        true);
+  }
+
+  private Optional<NearestKeyValue> getNearest(
+      final SegmentIdentifier segmentIdentifier,
+      final Predicate<Map.Entry<Bytes, Optional<byte[]>>> samePrefixPredicate,
+      final Predicate<Map.Entry<Bytes, Optional<byte[]>>> fallbackPredicate,
+      final boolean useMin)
+      throws StorageException {
 
     final Lock lock = rwLock.readLock();
     lock.lock();
     try {
-      // TODO: revisit this for sort performance
-      Comparator<Map.Entry<Bytes, Optional<byte[]>>> comparing =
-          Comparator.comparing(
-                  (Map.Entry<Bytes, Optional<byte[]>> a) -> a.getKey().commonPrefixLength(key))
-              .thenComparing(Map.Entry.comparingByKey());
-      return this.hashValueStore
-          .computeIfAbsent(segmentIdentifier, s -> newSegmentMap())
-          .entrySet()
-          .stream()
-          // only return keys equal to or less than
-          .filter(e -> e.getKey().compareTo(key) <= 0)
-          .sorted(comparing.reversed())
-          .findFirst()
-          .map(z -> new NearestKeyValue(z.getKey(), z.getValue()));
+      final Map<Bytes, Optional<byte[]>> segmentMap =
+          this.hashValueStore.computeIfAbsent(segmentIdentifier, s -> newSegmentMap());
+
+      // Define a function to find the nearest key-value pair based on a given predicate
+      final Function<Predicate<Map.Entry<Bytes, Optional<byte[]>>>, Optional<NearestKeyValue>>
+          findNearest =
+              (predicate) -> {
+                // Filter the segment map's entries based on the predicate
+                final Stream<Map.Entry<Bytes, Optional<byte[]>>> filteredStream =
+                    segmentMap.entrySet().stream().filter(predicate);
+                // Depending on the useMin flag, find either the minimum or maximum entry according
+                // to key order
+                final Optional<Map.Entry<Bytes, Optional<byte[]>>> sortedStream =
+                    useMin
+                        ? filteredStream.min(Map.Entry.comparingByKey())
+                        : filteredStream.max(Map.Entry.comparingByKey());
+                // Map the found entry to a NearestKeyValue object
+                return sortedStream.map(
+                    entry -> new NearestKeyValue(entry.getKey(), entry.getValue()));
+              };
```

Comment:
The only param from the method signature used in this function is useMin.  Can we define 2 static predicates, one for before and one for after, rather than creating a new function for every call to getNearest?    I would presume that allocating a new function for every call to getNearest is unnecessarily expensive

Reply:
we are also using segmentMap and this one is dynamic . not sure how you want to do, if you can give me an example to understand your idea

Reply:
it might or might not be more expensive  to create a function for every call versus just having a static function we pass parameters to.  It would be an interesting question to answer with profiling

Reply:
maybe @ahamlat can help for that ? if you think it will impact performance I will modify this part

Reply:
We can do micro-benchmarking with JMH on both implementations and enable Async profiler during the test. I just need to have an example of the usage of both methods, like a unit test.